Add new soil water modules

eloch216 opened the pull request:
Here I have started adding the new soil water modules to the BioCro module library. I'm marking this PR as a draft since there are still quite a few changes that need to be made.

So far, I have done the following:
1. Copied the modules into this repository. This required changing the `BioCroWater` namespace to `standardBML`. I also renamed `BioCroWater:soil_evaporation` to `BioCro:soil_evaporation2`, since there is already a `BioCro:soil_evaporation` module.
2. Added basic test cases for each  new module.
3. Added a new crop model called `soybean2` that uses the new modules. I also added a regression test for this model.

Future alterations should be made via pull requests that merge into the `add-new-soil-water-modules` branch.

The long term goal will be to remove the old soil water modules and completely replace them with the new ones.

I am going to start working on re-formatting the module files so they follow usual BioCro style.

eloch216 committed:
Copy over soil water modules

Changed the namespace to `standardBML`

eloch216 committed:
Add basic module tests

eloch216 committed:
Add new soybean model using new modules

eloch216 committed:
Run clang-format

eloch216 committed:
Do not include `Rinternals`

eloch216 committed:
Create soybean2.Rd

eloch216 committed:
Merge branch 'add-new-soil-water-modules' into restyle-new-soil-modules

eloch216 committed:
Don't make `using` directives at global scope

eloch216 committed:
Add header guards and make sure `includes` are necessary

eloch216 committed:
Use `.h` extension instead of  `.hpp`

leavesandlemmas commented on src/module_library/soil_water_flow_functions.h:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @brief Structure for storing the output from `infil` and `satflo`.
+ */
+struct infilWater_str {
```
It's my understanding that C-style arrays are strongly discouraged in C++, mostly for memory safety. `std::array` might be preferable?

I have started rewriting some of the canopy layer code that uses C-style arrays, namely `light_profile` in `sunML.h` and I think be simpler to have a `struct` that holds each of the variables which vary by layer, then use a container for that struct. Eg.

```cpp
// this 
 struct SoilLayer {
     double a;
     double b;
};

std::array<SoilLayer, 10> layers;

// instead of 
struct SoilLayers{
    std::array<double, 10> a;
    std::array<double, 10> b;
};
```

leavesandlemmas commented on src/module_library/soil_water_upflow.h:
```diff
@@ -0,0 +1,326 @@
+{
+    int nlayers = 6;
+
+    double soil_depth[] = {
```
I am willing to help implement any of my suggestions  of course

leavesandlemmas replied:
Is there a deadline to make changes? I can look at it today. But might not be able to finish until next week.

eloch216 committed:
Merge pull request #242 from biocro/restyle-new-soil-modules

Restyle new soil modules

eloch216 committed:
Remove unused argument

yufenghe22 commented on src/module_library/multi_layer_soil_profile.h:
```diff
@@ -0,0 +1,296 @@
+          surface_runoff{get_input(input_quantities, "surface_runoff")},
+          
+          // Parameters for layer 1
+          soil_depth_1{get_input(input_quantities, "soil_depth_1")},
```
Agreed. This is clumsy. I think vectorization would help.

yufenghe22 committed:
change kd to k_diffuse

eloch216 committed:
Try to document the albedo calculations

yufenghe22 committed:
update partitioning coefs

eloch216 committed:
Make code match documentation

eloch216 committed:
Add input checks

eloch216 committed:
Rewrite using piecewise method instead of clamping

yufenghe22 committed:
small updates on soybean2.R

yufenghe22 committed:
remove hard-coded constants

yufenghe22 committed:
change time_factor to days_stage2

yufenghe22 committed:
add hours_per_day

yufenghe22 committed:
organize hard-coded constants

yufenghe22 committed:
add some comments

yufenghe22 committed:
move soil_evaporation_alpha

yufenghe22 committed:
Revert "small updates on soybean2.R"

This reverts commit 8d34640d472946efd2696ef00b32c42341b06b20.

yufenghe22 committed:
Revert "small updates on soybean2.R"

This reverts commit 8d34640d472946efd2696ef00b32c42341b06b20.

yufenghe22 committed:
update test files for soybean2

yufenghe22 committed:
update test files for soybean2

yufenghe22 committed:
require this module to use Euler solver

yufenghe22 committed:
minor changes

yufenghe22 committed:
rearrange soil_evaporation_alpha

eloch216 committed:
Add another reference

eloch216 committed:
Merge pull request #244 from biocro/soil-albedo

Soil albedo

eloch216 committed:
Merge pull request #249 from biocro/fix_soil_evaporation

Fix soil evaporation

eloch216 committed:
Merge remote-tracking branch 'origin/add-new-soil-water-modules' into update-parameters-new-water

eloch216 committed:
Update test cases

eloch216 committed:
Fix tiny mistakes made while addressing merge conflicts

yufenghe22 committed:
Merge pull request #246 from biocro/update-parameters-new-water

Update partitioning parameters with new water

leavesandlemmas commented:
> The two state variables are `sw_avail` and `soil_water_content`, which are both in hourly step. The calculation is through layers, not through time. Is it already compatible with non-euler solver?

Looking at these two lines like seems like an euler time integration step is being performed. If `upflow` has units cm/hr, then it's like the rate at which water moves from one layer to the next.

```cpp
 upflow[l] = dbar * grad / ((soil_depth[l] + soil_depth[m]) * 0.5);  // to cm/hr
 /* 

snip 

*/     
                swtemp[l] = swtemp[l] + upflow[l] / soil_depth[l];
```

Let $x_i$ be the water content of  $i$-th soil layer. Let $F_i$ be the upflow rate.  Then $\dot{x}_i$ has unit dimensions per time, whereas if $F_i$ has units cm / hr, then dividing by soil depth would have the correct units to be $\dot{x}_i$.  I am not sure dividing by soil depth makes sense to me, but that's besides the point.  If I assume the model's logic is sound, and we want this code to return a rate of change in soil water content.

yufenghe22 commented:
> But why "soil depth" and not "soil layer thickness"

IT IS indeed thickness! I, many years ago, suggested that the name should be changed. The fortran code is slighly better named `DLAYR`,
`! DLAYR(L)    Thickness of soil layer L (cm)`
But I think `thickness` should be the one to use, even for variable use.

leavesandlemmas commented:
> > But why "soil depth" and not "soil layer thickness"
> 
> IT IS indeed thickness! I, many years ago, suggested that the name should be changed. The fortran code is slighly better named `DLAYR`, `! DLAYR(L) Thickness of soil layer L (cm)` But I think `thickness` should be the one to use, even for variable use.

Ohhh, okay. Im going to change the name then because that is even more confusing that the names like `esw`

leavesandlemmas commented:
@yufenghe22 When you have time, would you verify that this code has the same meaning?

```cpp
    // formerly esw
    double extractable_water(const soil_layer& layer) const {
        return layer.field_capacity - layer.wilting_point;
    }
    
    // formerly thet1 and thet2
    // I think this should have a better name. Excess water conten?
    double theta(const soil_layer& layer) const {
        double esw = extractable_water(layer);
        double th = std::min(layer.water_content - layer.wilting_point, esw);
        return std::max(0.0, th);
    }


    double average_diffusivity(const soil_layer& layer, const soil_layer& next_layer) const {
        
        double avg_theta  = (theta(layer) + theta(next_layer))/2;
        
        constexpr double magic_num_1 = 35.4; 
        double u = magic_num_1  * avg_theta;

        constexpr double min_diffusivity = 0.88; 
        double out = min_diffusivity * std::exp(u); 
        
        constexpr double max_diffusivity = 100;
        return std::min(out, max_diffusivity); 

    }
    
    // is this a gradient ? spatial derivative?
    double grad(const soil_layer& layer, const soil_layer& next_layer) const {
             
        double a = theta(next_layer) / extractable_water(next_layer) - theta(layer) / extractable_water(layer);
        double w = layer.depth / (layer.depth + next_layer.depth);
        double weighted_extractable_water = w * extractable_water(layer) + extractable_water(next_layer) * (1 - w);
        return a * weighted_extractable_water;
    }

    double upflow(const soil_layer& layer, const soil_layer& next_layer) const {
        double avg_thickness = (layer.thickness + next_layer.thickness) * 0.5;
        return average_diffusivity(layer, next_layer) * grad(layer, next_layer) / avg_thickness;  // cm / hr    
    }
```

eloch216 committed:
Add documentation for potential_soil_evaporation

And incomplete documentation for reference_evapotranspiration

eloch216 committed:
Reorganize and add more documentation

All tests are still passing

eloch216 committed:
Remove extra LAI calculations

eloch216 committed:
Remove potential_evapotranspiration

It is not used for anything

eloch216 committed:
Make variable names match documentation

eloch216 committed:
Merge branch 'develop' into add-new-soil-water-modules

eloch216 committed:
Merge branch 'add-new-soil-water-modules' into potential-soil-evaporation

eloch216 committed:
Change basal crop coefficient parameter values

And un-hard-code the value of kcbmin

eloch216 committed:
Specify const for safety

eloch216 committed:
Merge pull request #267 from biocro/potential-soil-evaporation

Potential soil evaporation

eloch216 committed:
Use consistent atmospheric pressure

Changes to stored soybean2 model results are very small

eloch216 committed:
Use consistent atmospheric transmittance

Changes to stored soybean2 model results are very small

eloch216 committed:
Use instantaneous extraterrestrial radiation

Also fix estimate of radiation at Earth's surface so it includes NIR energy in addition to PAR.

Changes to stored soybean2 regression test are pretty small

eloch216 committed:
Use consistent e_sat

And don't worry about low RH values, since the "correction" doesn't make sense here (because we don't know the daily minimum temperature)

eloch216 committed:
Stop fixing windspeed_height to 2 m

Changes to regression test are very small

eloch216 committed:
Use correct value of sigma when calculating rnl

There is a fairly significant decrease in biomass after this change. The previous value of sigma was given on a per day basis, so it was numerically larger. So now we have a smaller estimated value of longwave energy losses, causing an increase in the energy available to drive evapotranspiration, and hence higher soil evaporation rates.

eloch216 committed:
Account for soil heat flux

Changes to regression test data are very small

eloch216 committed:
Use correct Cd values

Changes to regression test are very small

eloch216 committed:
Allow negative reference ET

As suggested by ASCE (2005). Changes to stored biomass values are small.

eloch216 committed:
Use fractional_doy to be more continuous

eloch216 committed:
Alphabetize

eloch216 committed:
Add more documentation

eloch216 committed:
Be more careful with transmission coefficients

Changes to regression tests are very small

eloch216 committed:
Try to improve explanation for transmittance

eloch216 committed:
Use reference albedo

Very little change to stored regression data

eloch216 committed:
Fix long lines

eloch216 committed:
Try to improve documentation a little more

eloch216 committed:
Merge pull request #270 from biocro/fix-reference-ET

Fix reference ET

yufenghe22 committed:
add dynamic rooting

yufenghe22 committed:
fix some rates

yufenghe22 committed:
calibrate wiltp and fieldc

eloch216 committed:
Merge branch 'develop' into update-soil-water-from-develop

eloch216 committed:
Update DESCRIPTION

eloch216 committed:
Update soybean2 model definition and tests

eloch216 committed:
Merge pull request #279 from biocro/update-soil-water-from-develop

Update soil water from develop

yufenghe22 committed:
add soil_water_dynamic_rooting.h

yufenghe22 committed:
change comments

yufenghe22 committed:
update comment

yufenghe22 committed:
update test files

yufenghe22 committed:
clang format

yufenghe22 committed:
add test file

yufenghe22 committed:
update units

yufenghe22 committed:
clang-format

yufenghe22 committed:
include units in return

yufenghe22 committed:
Merge pull request #280 from biocro/fix_root_uptake

Fix root uptake

yufenghe22 committed:
go back to the values in Campbell and Norman

yufenghe22 committed:
change layer soil types

eloch216 committed:
Merge branch 'develop' into add-new-soil-water-modules

yufenghe22 committed:
add a new bb water stress modifier function

yufenghe22 committed:
cap the changes on b0

yufenghe22 committed:
use swc as the driving variable

yufenghe22 committed:
change default water stress module

yufenghe22 committed:
use REW as predictor

yufenghe22 committed:
do not diff StomataWS on b0 and b1

yufenghe22 commented:
What other things need to checked for this PR? Sorry it's been a while, and this PR is already quite chunky. I think we should merge this, keep the `soybean2`, and make new PR for other changes like officially making `soybean2` the `soybean`.

eloch216 commented:
> What other things need to checked for this PR? Sorry it's been a while, and this PR is already quite chunky.

There are still a large number of functions and modules here that have essentially no documentation. Examples of files with insufficient documentation include (but are not limited to) `surface_excess_water.h`, `soil_water_uptake.h`, `soil_water_upflow.h`, `soil_water_tiledrain.h`, `soil_water_flow_functions.cpp`, etc. I have gone through some functions, especially in `soil_evaporation_functions.h`, but there is still a lot more work to do.

This is tedious but important. If there is no clear indication for the source of an equation, it is difficult to tell if it is being used properly. Almost everything I have looked at so far has had some kind of mistake, often related to an assumed time step. Because of this, I don't really trust the code that hasn't been documented yet. And I don't want to add more undocumented code into BioCro.

> I think we should merge this, keep the `soybean2`, and make new PR for other changes like officially making `soybean2` the `soybean`.

I don't think we should merge this yet because the modules are not ready yet. If we merge it into `develop`, that would prevent us from making any new BioCro releases until this code is done. Otherwise, we would be forced to release modules that are not finalized.

I don't think `soybean2` is ready to become the new `soybean`, either. Besides the documentation issues, there is also the Euler solver issue. As a reminder, one of the modules -- `soil_evaporation2` -- requires a fixed-step Euler solver with a time step of 1 hour. It would be best to either replace this with a different model, or figure out how to modify it so it does not need to keep track of how long the soil has been in "Stage 2 evaporation."

yufenghe22 commented:
> Examples of files with insufficient documentation include (but are not limited to) `surface_excess_water.h`, `soil_water_uptake.h`, `soil_water_upflow.h`, `soil_water_tiledrain.h`, `soil_water_flow_functions.cpp`, etc.

Isn't `soil_water_uptake.h` documented properly? I merged a PR for this.

Would adding the sources of where these equations come from be sufficient? Most of them, probably all of them are from DSSAT's fortran code. Accepting DSSAT's codes as-is is not a bad idea, imo, especially when the prediction on soil water dynamics checks out. I agree some of the codes are not compatible with BioCro, such as those time step related constants. But again, some of them may have small impact on the prediction of soil water content results.

The bottom line is, we have a wrong soil water model _v.s_ a working soil water model that may have errors. It seems the latter is still a better choice.

> I don't think `soybean2` is ready to become the new `soybean`, either.

That's what I mean. We keep `soybean2`. Other changes can be made in new PRs.

> If we merge it into `develop`, that would prevent us from making any new BioCro releases until this code is done.

I'm not familiar with the standard of releases, but if having `soybean2` in the model prevents that, I guess that's a reason to not merge it.

eloch216 commented:
> The bottom line is, we have a wrong soil water model _v.s_ a working soil water model that may have errors. It seems the latter is still a better choice.

The other bottom line is that we previously had poorly documented soil water code, and we now have _different_ poorly documented soil water code. So I would argue that in its current state, this PR is only a half improvement.

> I agree some of the codes are not compatible with BioCro, such as those time step related constants. But again, some of them may have small impact on the prediction of soil water content results.

Surely some of the mistakes will only have a small impact. But others could have a large impact, and there is no way to tell until we check through all the code.

> Would adding the sources of where these equations come from be sufficient? Most of them, probably all of them are from DSSAT's fortran code.

A model is not just lines of code -- it is a set of equations designed to represent some aspect of a physical system that were derived under certain assumptions. The purpose of documentation is so that a user or developer can understand these aspects of a model, enabling them to judge whether it is a good choice for a particular application.

The underlying assumptions of a model are typically only explained in textbooks or journal articles. So, ideally, our documentation should indicate where to find every BioCro equation in one of these sources. Sometimes that is not possible, but in such cases, we should provide a quick justification of our own to explain why an equation was used. If we cannot provide a textbook/journal reference or our own justification for an equation, we probably shouldn't be using it.

Our legacy code does not always meet these documentation requirements. But we should try to make sure any new code is sufficiently documented.

It's possible that citing DSSAT could be justified in some cases, but there are several reasons for citing DSSAT on its own is not usually sufficient:
1. DSSAT includes dozens of files and thousands of lines of code so it is difficult to find a particular equation within it.
2. Even if a more specific reference is given to a particular DSSAT subroutine and/or file, it is not always easy to identify an equation because we may use different variable names.
3. If the DSSAT code is updated, our code will not be automatically updated to follow the latest version. Conceptually, our code is based on the same references as the current version of DSSAT, but not tied to DSSAT itself.
4. Much of DSSAT is not documented well, so referring to DSSAT might not provide any insight into a model's justification or underlying assumptions.
5. Even when DSSAT does provide clear references to a textbook or journal article, we could just as easily provide the references on our own so an interested user doesn't have to follow a sequence of nested references to find the original source.

> Isn't `soil_water_uptake.h` documented properly? I merged a PR for this.

For this particular module, I have a few questions that don't seem to be addressed by the documentation:
1. Is this based off DSSAT or something else in the literature? Or is it brand new? I know you developed part of it, but did you develop all of it?
2. Does this assume an Euler solver with a time step of 1 hour? It seems to compare `uptake` (a rate with units of `Mg / ha / hr`) against `max_water` (a soil water content with units of `Mg / ha`). This seems to imply a time step of 1 hour.
3. One of the explanatory comments seems to contradict the code: `// uptake in layer = ET*(depth of layer/total depth of root zone)`. It looks like uptake is calculated using the fraction of root mass found in each layer, not the depth of each layer?

yufenghe22 commented:
> 1. Is this based off DSSAT or something else in the literature? Or is it brand new? I know you developed part of it, but did you develop all of it?

I built all of it. See also my [response in the PR](https://github.com/biocro/biocro/pull/280#discussion_r2913509127). I will clarify it in the module description.

> 2\. Does this assume an Euler solver with a time step of 1 hour? It seems to compare `uptake` (a rate with units of `Mg / ha / hr`) against `max_water` (a soil water content with units of `Mg / ha`). This seems to imply a time step of 1 hour.

No. It does not imply 1 hour time step. `max_water` does not have a time part in its unit. I think whatever the unit of `uptake` has, the module is okay with it.

Wait. If the step is too large (like daily), there would be an issue. I guess it won't be a problem for non-euler solvers. Are they always with steps < 1 hour? I have a way to fix this. Do we have a "delta time" variable? Is the `timestep` from the input `hourly`? I mean, for non-euler solvers, there should be an auto-adjusted time step variable.

> 3\. One of the explanatory comments seems to contradict the code: `// uptake in layer = ET*(depth of layer/total depth of root zone)`

This is a legacy comment. I'll remove it.

leavesandlemmas commented:
> Do we have a "delta time" variable? Is the `timestep` from the input `hourly`? I mean, for non-euler solvers, there should be an auto-adjusted time step variable.

There is not really a single `Δt`.  As I have discovered the hard way, each ODE method works differently and there's not a consistent interface.  If I understand them correctly, `run_biocro` checks if the `time` variable is spaced by `timetstep` and all differential quantities are multiplied by `timestep` so it sets the units of time implicitly but only  `homemade_euler` treats `timestep` as its step size. 

I believe `boost_euler` on the other hand users `output_step_size` as its step size which can be larger or smaller than `timestep` as does `boost_rk4`  and `boost_rkck54` but  `boost_rosenbrock`  picks its own step size based on the error tolerances its given. AFAIK, there isn't any way to see what step size it chooses, and it definitely chooses poorly sometimes.  

`boost_rkck54` is adaptive but I think it won't take a bigger step than `output_step_size` but might break in interval into several subintervals if it estimates the error is too large.   But `timestep` doesn't affect these solvers at all.  So there is an auto-adjusted time step variable, but it isn't visible to biocro modules. 

>  Wait. If the step is too large (like daily), there would be an issue. I guess it won't be a problem for non-euler solvers. Are they always with steps < 1 hour? 

No, not necessarily. With the exception of the `homemade_euler`, the solvers may take any step size regardless.  

Normally, this is what you'd want from a ODE integrator. It assumes a vector field `f(t, y) = dy/dt` that is defined for all times, so it can decide how big a step to take, so there's nothing special about integrating for < 1 hour or > 1 hour.

eloch216 commented:
> > 2. Does this assume an Euler solver with a time step of 1 hour? It seems to compare `uptake` (a rate with units of `Mg / ha / hr`) against `max_water` (a soil water content with units of `Mg / ha`). This seems to imply a time step of 1 hour.
> 
> No. It does not imply 1 hour time step. `max_water` does not have a time part in its unit. I think whatever the unit of `uptake` has, the module is okay with it.

I am looking at this line in particular:

https://github.com/biocro/biocro/blob/ca094136b3581b9005220dfd01d59fa9c8a5fcf7/src/module_library/soil_water_uptake.h#L254

In my understanding, `requested_uptake` has the same units as `base_request`, which in turn has the same units as `canopy_transpiration_rate`: `Mg / ha / hr`. But `max_negative_uptake` has the same units as `max_water`: `Mg / ha`. Two quantities with different units cannot be compared. This would be like asking whether `10 m / s` is greater than `9 m`. When I have seen this kind of unit issue in the past, it has typically occurred because of an implied `delta_time` somewhere. For example, `requested_uptake * one_hour` would then have units of `Mg / ha` and would be comparable to `max_negative_uptake`. If a particular value of a time interval is being assumed, then this assumption can only be met when using a fixed-step Euler solver.

> I have a way to fix this. Do we have a "delta time" variable? Is the `timestep` from the input `hourly`? I mean, for non-euler solvers, there should be an auto-adjusted time step variable.

No, in BioCro there is no way for a module to know the time step being taken by the ODE solver. This is intentional. A core part of the BioCro design philosophy is that model equations should always be separate from the solution algorithm used to solve them. Mixing the equations and the solver together produces code and models that are difficult to understand. As Scott has described above, the time step is part of the solver, not of the model equations.

justinmcgrath commented:
> 1 DSSAT includes dozens of files and thousands of lines of code so it is difficult to find a particular equation within it.
> 
> 2 Even if a more specific reference is given to a particular DSSAT subroutine and/or file, it is not always easy to identify an equation because we may use different variable names.
> 
> 3 If the DSSAT code is updated, our code will not be automatically updated to follow the latest version. Conceptually, our code is based on the same references as the current version of DSSAT, but not tied to DSSAT itself.
> 
> 4 Much of DSSAT is not documented well, so referring to DSSAT might not provide any insight into a model's justification or underlying assumptions.
> 
> 5 Even when DSSAT does provide clear references to a textbook or journal article, we could just as easily provide the references on our own so an interested user doesn't have to follow a sequence of nested references to find the original source.

Here's some defense of citing DSSAT.
Points 1 and 3 are addressed by giving the specific version of DSSAT and the file name and line numbers for the code.
Point 2 is almost always the case; when citing original sources we usually use different names too.
For 5, when you're finding the specific part of DSSAT to cite, you'd see these, and could include them or cite only them.

Point 4 is a problem, and it would be nice to not use things that aren't described.

We need to get this out there, and unfortunately I think that will require relaxing standards to some degree, so good faith efforts to cite parts of DSSAT may be as good as it gets. However, getting the units and the treatment of the time step right are critical. I see a lot where those are not stated. Can that be prioritized? As a workflow note, it's much easier to write those things when writing the code than it is to go back later.

leighmatth commented:
I understand the concerns about merging, but this is significantly better than what is currently in biocro and we do need to get a version of this available. What would be the must-haves for now and what could be worked on later? I think its okay to have this right now only be used with the Euler method, and we can work on making it workable with the other integrators in the future.

Another option would be to keep this as separate module library for now, but maybe refer to it in the main biocro readme.


> As a workflow note, it's much easier to write those things when writing the code than it is to go back later.

Agreed, and I brought this up, but unfortunately the person who wrote this code left before they had to be the one to deal with the challenge we are now facing.

eloch216 commented:
> We need to get this out there, and unfortunately I think that will require relaxing standards to some degree, so good faith efforts to cite parts of DSSAT may be as good as it gets. However, getting the units and the treatment of the time step right are critical. I see a lot where those are not stated. Can that be prioritized? As a workflow note, it's much easier to write those things when writing the code than it is to go back later.

Point taken, I am probably trying to be too strict about this. I think documenting the units would be a reasonable goal to aim for. Part of doing this may require tracking down original references, and if/when that happens, they can also be added to the documentation.

yufenghe22 committed:
make input references

yufenghe22 committed:
make linear fitted coefs inputs

yufenghe22 committed:
add units and clang-format

yufenghe22 committed:
add test files

yufenghe22 committed:
add data and scripts to inst

yufenghe22 committed:
update test

yufenghe22 committed:
add some documentations on soil type

dilkaran committed:
Add units to soil_water_downflow.h

eloch216 committed:
Re-format soil table

dilkaran committed:
Units added to soil_water_tiledrain.h

eloch216 committed:
Update script and documentation

dilkaran committed:
Units added to soil_water_tiledrain.h

eloch216 committed:
Remove old linear module

eloch216 committed:
Rename v2 of stomata module

eloch216 committed:
Add more documentation

eloch216 committed:
Update stomata_water_stress_linear.h

yufenghe22 committed:
Merge pull request #287 from biocro/change_soil_types

Change soil types

yufenghe22 committed:
use csv instead of rds

yufenghe22 committed:
update file read-in function

dilkaran committed:
Simplify units: cm^3 [water] / cm^3 [soil] -> cm^3 / cm^3

dilkaran committed:
Simplify units: cm^3 [water] / cm^3 [soil] -> cm^3 / cm^3

eloch216 committed:
Add a lot of documentation, including units

dilkaran committed:
Units added to soil_water_upflow.h

eloch216 committed:
Remove many unused inputs

eloch216 committed:
Fix typo

eloch216 committed:
Alphabetize quantities and rename module

yufenghe22 committed:
add units to multi_layer_soil_profile.h

eloch216 committed:
Merge pull request #289 from biocro/updates-to-water-stress-func

Updates to water stress func

yufenghe22 committed:
Merge branch 'add-new-soil-water-modules' into use_new_water_stress_function

eloch216 committed:
Change soil evaporation rate units

This is necessary for consistency with other modules

yufenghe22 committed:
change the input E's unit from mm/hr to Mg/ha/hr

yufenghe22 committed:
update tests

yufenghe22 committed:
Merge branch 'soil-evaporation-units' into units_multi_layer_soil_profile

yufenghe22 committed:
add units and timestep

yufenghe22 committed:
Merge branch 'units_multi_layer_soil_profile' of github.com:biocro/biocro into units_multi_layer_soil_profile

yufenghe22 committed:
update soybean2 test file

yufenghe22 committed:
Merge pull request #294 from biocro/units_multi_layer_soil_profile

add Units for multi layer soil profile

eloch216 committed:
Merge pull request #292 from biocro/soil-evaporation-units

Soil evaporation units

dilkaran committed:
TODO resolved | Long comments following units are removed

dilkaran committed:
Unused code (commented out) removed

dilkaran committed:
Surface runoff var removed

dilkaran committed:
Merge pull request #291 from biocro/soil-water-upflow-units

Units added to soil_water_upflow.h

eloch216 committed:
Documentation for `infil`

yufenghe22 committed:
add units to multilayer_soil_profile_avg.h

yufenghe22 committed:
add units to dynamic_rooting

yufenghe22 committed:
Update comments in soil_water_dynamic_rooting.h

yufenghe22 committed:
add units to soil_water_uptake.h

dilkaran committed:
Removed TODO and longer sentences

dilkaran committed:
Removed variable total_tile_flow

dilkaran committed:
removed long comment

yufenghe22 committed:
define var types

dilkaran committed:
Merge pull request #288 from biocro/soil_water_downflow-units

Units added to soil_water_downflow.h

leavesandlemmas committed:
updated units; removed unused code

yufenghe22 committed:
updated some formats

yufenghe22 committed:
Merge pull request #297 from biocro/units_soil_water_uptake

add units to soil_water_uptake.h

yufenghe22 committed:
Merge branch 'add-new-soil-water-modules' into use_new_water_stress_function

dilkaran committed:
renamed cumulative_tile_flow to tile_flow_rate. Updated stored soybean2 regression data to reflect variable name change and removal of variable total_tile_flow

dilkaran committed:
renamed cumulative_tile_flow to tile_flow_rate

yufenghe22 committed:
Merge pull request #295 from biocro/units_multi_soil_avg

add units to multilayer_soil_profile_avg.h

eloch216 committed:
Documentation for `satflo`

eloch216 committed:
Remove unused input argument

eloch216 committed:
Move timestep and potential infiltration

eloch216 committed:
Use better names and more consistent units

leavesandlemmas committed:
requested updates

leavesandlemmas committed:
Update module_library.cpp

leavesandlemmas committed:
fix tests

leavesandlemmas committed:
Added refs

yufenghe22 committed:
Merge branch 'add-new-soil-water-modules' into soil-water-tiledrain-units

dilkaran committed:
Merge pull request #290 from biocro/soil-water-tiledrain-units

Units added to soil_water_tiledrain.h

eloch216 committed:
Use `m^3 / m^3` instead of `dimensionless`

eloch216 committed:
Remove some magic numbers

eloch216 committed:
Remove unused variable

eloch216 committed:
Add another reference

eloch216 committed:
Fix compiler error and update tests

eloch216 committed:
Merge pull request #302 from biocro/fix-compiler-error

Fix compiler error and update tests

eloch216 committed:
Merge branch 'add-new-soil-water-modules' into infil-and-satflo-units

eloch216 committed:
Resolve merge conflict

eloch216 committed:
Merge pull request #300 from biocro/infil-and-satflo-units

Infil and satflo units

yufenghe22 committed:
use rsdf and min_root_depth

yufenghe22 committed:
add rsdf

yufenghe22 committed:
update module info

yufenghe22 committed:
update tests

yufenghe22 committed:
Merge branch 'add-new-soil-water-modules' into units_dynamic_rooting

leavesandlemmas committed:
hardcoding timestep

yufenghe22 committed:
use cm for min_root_depth

yufenghe22 committed:
update test file

eloch216 committed:
Document units in `up_flow`

eloch216 committed:
clang-format

eloch216 committed:
Fix if-else block formatting

eloch216 committed:
Document units in `tile_flow`

eloch216 committed:
Use consistent names

yufenghe22 committed:
Merge pull request #296 from biocro/units_dynamic_rooting

add units to dynamic_rooting

eloch216 committed:
Add unit and reference

eloch216 committed:
Fix typo

eloch216 committed:
Merge branch 'add-new-soil-water-modules' into upflow-tileflow-units

eloch216 committed:
Address merge conflict

eloch216 committed:
Merge pull request #303 from biocro/upflow-tileflow-units

Upflow tileflow units

eloch216 committed:
Merge pull request #298 from biocro/scotts-assigned-soil-water-mods

updated units; removed unused code

yufenghe22 committed:
Merge branch 'add-new-soil-water-modules' into use_new_water_stress_function

eloch216 committed:
Run clang-format

And a few other minor style changes

eloch216 committed:
Remove unneeded variable and static_cast

yufenghe22 committed:
update StomataWS coefs

yufenghe22 committed:
update tests

eloch216 committed:
Fix unit parsing issues and standardize some units

eloch216 committed:
Revert unrelated change

eloch216 committed:
Fix test case

yufenghe22 committed:
Merge pull request #284 from biocro/use_new_water_stress_function

use a new ball-berry water stress function

eloch216 committed:
Merge branch 'add-new-soil-water-modules' of https://github.com/biocro/biocro into add-new-soil-water-modules

eloch216 committed:
Update NEWS.md

yufenghe22 commented:
> I think this is finally ready to go. Thanks to @dilkaran and @leavesandlemmas for helping @yufenghe22 and me with the unit documentation.
> 
> I just pushed a few more commits where I applied `clang-format` and standardized some units. I have also confirmed that there are no unit conflicts between the different modules.
> 
> There are probably still opportunities to improve these modules more. But including units and references was a major, and helpful, step. Other improvements can be made at later times.

Thanks very much for coordinating the tasks! it's great to have a version deposited.
